python/ml4ir/base/model/calibrations/temperature_scaling.py

```diff
@@ -0,0 +1,247 @@
+    """
+    # convert to TF.Tensor
+    logits_tensor = tf.constant(logits, name='logits_test', dtype=tf.float32)
+    labels_tensor = tf.constant(labels, name='labels_test', dtype=tf.int32)
+
+    if temperature:
+        logits_tensor = tf.divide(logits_tensor, temperature)
+    nll = tf.reduce_mean(
+        tf.nn.sparse_softmax_cross_entropy_with_logits(labels=labels_tensor, logits=logits_tensor))
+
+    # getting the softmaxes using the model final activation function
+    softmaxes = scorer.final_activation_op(logits_tensor, {})
+    return accuracy(softmaxes, labels_tensor), tf.identity(nll), softmaxes
+
+
+def get_logits_labels(model: tf.keras.Model, evaluation_set: tf.data.TFRecordDataset):
+    """
+    predicts model output on the given evaluation set
+    Returns:
+        model output: numpy.array
+        labels: numpy.array
+    """
+    logits_nps = model.predict(evaluation_set).squeeze()
+    labels_nps = np.concatenate(list(map(lambda x: x[1], evaluation_set))).squeeze().argmax(axis=-1)
+    return logits_nps, labels_nps
+
+
+def temperature_scale(model: tf.keras.Model, scorer: ScorerBase, dataset: RelevanceDataset,
+                      logger: Logger,
+                      logs_dir_local: str, temperature_init: float):
+    """
+    learns a temperature parameter using Temperature Scaling (TS) technique on the validation set.
+    It, then, computes the probability scores of the test set with and without TS and writes them
+    in a .zip file.
+    Parameters:
+        model : tf.keras.Model object to be used for temperature scaling
+        scorer:  ScorerBase object of the RelevanceModel
+        dataset: RelevanceDataset to be used for temperature scaling.
+        logger : Logger
+        logs_dir_local: path to save the TS scores
+        temperature_init: temperature initial value
+    """
+    # get an intermediate model with logits as output --> MUST NOT BE AN ACTIVATION (e.g.
+    # SOFTMAX) LAYER
+    intermediate_model = get_intermediate_model(model, scorer)
+
+    start = tf.Variable(initial_value=[temperature_init], shape=(1), dtype=tf.float32, name="temp",
+                        trainable=True)
+
+    def nll_loss_fn(temperature):
+        """
+        computes Negative Log Likelihood loss by applying temperature scaling.
+        Parameters:
+            temperature: TF.Tensor,  a  parameter of size=(1) to be trained.
+        """
+        logits_tensor = tf.constant(logits_nps, name='logits_tensor', dtype=tf.float32)
+        labels_tensor = tf.constant(labels_nps, name='labels_tensor', dtype=tf.int32)
+        logits_w_temp = tf.divide(logits_tensor, temperature)
+
+        return tf.reduce_mean(
+            tf.nn.sparse_softmax_cross_entropy_with_logits(labels=labels_tensor,
+                                                           logits=logits_w_temp))
+
+    @tf.function
+    def nll_with_lbfgs():
+        return tfp.optimizer.lbfgs_minimize(make_val_and_grad_fn(nll_loss_fn),
```

Comment:
@marjanhs Can you add description for why we are using this here instead of say doing SGD? A reference should work too. More as documentation for future.

Reply:
I have added the reference on top of the page:
 https://github.com/salesforce/ml4ir/pull/115/files#diff-5e55239ed1f122b6b40beccad8e8301108acd2e382474342814c595590f776dcR177
(the implementations are inspired by the reference)

Reply:
but added here , too.